src/ripple/consensus/Validations.h

```diff
@@ -0,0 +1,690 @@
+        @param cutoffBefore Ignore ledgers with sequence number before this
+    */
+    hash_map<LedgerID, ValidationCounts>
+    currentTrustedDistribution(
+        LedgerID const& currentLedger,
+        LedgerID const& priorLedger,
+        std::uint32_t cutoffBefore)
+    {
+        bool const valCurrentLedger = currentLedger != beast::zero;
+        bool const valPriorLedger = priorLedger != beast::zero;
+
+        hash_map<LedgerID, ValidationCounts> ret;
+
+        current(
+            stalePolicy_.now(),
+            // The number of validations does not correspond to the number of
+            // distinct ledgerIDs so we do not call reserve on ret.
+            [&](std::size_t) {},
+            [&](NodeKey const&, Validation const& v) {
+
+                if (!v.trusted())
+                    return;
+
+                std::uint32_t const seq = v.seq();
+                if ((seq == 0) || (seq >= cutoffBefore))
+                {
+                    // contains a live record
+                    bool countPreferred =
+                        valCurrentLedger && (v.ledgerID() == currentLedger);
+
+                    if (!countPreferred &&  // allow up to one ledger slip in
+                                            // either direction
+                        ((valCurrentLedger &&
+                          v.isPreviousLedgerID(currentLedger)) ||
+                         (valPriorLedger && (v.ledgerID() == priorLedger))))
+                    {
+                        countPreferred = true;
+                        JLOG(j_.trace()) << "Counting for " << currentLedger
+                                         << " not " << v.ledgerID();
+                    }
+
+                    ValidationCounts& p =
+                        countPreferred ? ret[currentLedger] : ret[v.ledgerID()];
+                    ++(p.count);
+
+                    NodeID const ni = v.nodeID();
+                    if (ni > p.highNode)
+                        p.highNode = ni;
+                }
+            });
+
+        return ret;
+    }
+
+    /** Count the number of current trusted validators working on the next
+        ledger.
+
+        Counts the number of current trusted validations that replaced the
+        provided ledger.  Does not check or update staleness of the validations.
+
+        @param ledgerID The identifier of the preceding ledger of interest
+        @return The number of current trusted validators with ledgerID as the
+                prior ledger.
+    */
+    std::size_t
+    getNodesAfter(LedgerID const& ledgerID)
+    {
+        std::size_t count = 0;
+
+        // Historically this did not not check for stale validations
+        // That may not be important, but this preserves the behavior
+        current(
+            boost::none,
+            [&](std::size_t) {}, // nothing to reserve
+            [&](NodeKey const&, Validation const& v) {
+                if (v.trusted() && v.isPreviousLedgerID(ledgerID))
```

Comment:
Yeah, that would change the assumption about what's being counted in `currentTrustedDistrubution`. The problem is that the count under the current assumption (same ledger chain) can be inaccurate if we don't set the previous ledger id for a validation. It seems like we have the information to determine that a lone validation validates a ledger id whose sequence is immediately after a given sequence. It would just be a matter of finding the least messy way of figuring that out.